### ISSUE TITLE
GitHub reporter

### ISSUE THREAD

**Issue by hugo-vrijswijk** [stryker-mutator/stryker4s#594]
GitHub has something called 'check-runs' they allow adding annotations to a 'check'. Checks are what GitHub Actions create. These annotations show up at a line of code. A good example is the [eslint-action](https://github.com/gimenete/eslint-action) which lints code and displays error messages directly in GitHub. More info about checks can be found [here](https://docs.github.com/en/free-pro-team@latest/rest/reference/checks#runs) or [here](https://developer.github.com/changes/2018-05-07-new-checks-api-public-beta/).  Annotations can be added to a new check created by the action, or an existing one that is currently running like the CI action. It looks a little like this:

![eslint-action](https://raw.githubusercontent.com/gimenete/eslint-action/master/screenshots/annotation.png)

We could add something similar for Stryker4s by adding a GitHub reporter. This reporter could annotate survived mutants at their position in the source code on GitHub so people can view their mutation testing results in GitHub. This would also work very well in combination with a diff-runner (not supported yet by Stryker4s).

Some things that need discussion:

- What should survived mutants be reported as? I don't think it should be an 'error' because then the user will always have a failed check unless they have a score of 100% which is the wrong message to send IMO
- Should we report other mutant statuses? It might become a little crowded if we report everything
- Another option would be a stryker-report-action that could be run after mutation testing. The action could read the JSON report and send the annotations. Upside is that it would work for all Strykers (or other mutation testing framework). Downside is it would not report the results in realtime, and would require the main CI to be in GitHub actions already (which wouldn't be required if the Stryker4s reporter creates a new check)

@stryker-mutator/stryker-js-and-friends @stryker-mutator/stryker-net do you have any thoughts? Nothing about this is Stryker4s specific

### Technical solution

We can add a [new 'reporter' type](https://github.com/stryker-mutator/stryker4s/blob/2d4744118bf05368342c48b5a7dcb930800f59a7/core/src/main/scala/stryker4s/run/Stryker4sRunner.scala#L38-L48) in the [configuration](https://github.com/stryker-mutator/stryker4s/blob/master/core/src/main/scala/stryker4s/config/ReporterType.scala). There is already a reporter that does http requests: the [dashboard reporter](https://github.com/stryker-mutator/stryker4s/blob/master/core/src/main/scala/stryker4s/report/DashboardReporter.scala). Collecting the environment variables and other configuration should probably be done in a [similar way](https://github.com/stryker-mutator/stryker4s/blob/master/core/src/main/scala/stryker4s/report/dashboard/DashboardConfigProvider.scala). The new github reporter could open a websocket connection using [sttp fs2 websockets](https://sttp.softwaremill.com/en/latest/backends/fs2.html#websockets) to stream the data to GitHub, or collect, chunk, and send results in batches via HTTP requests. Ideally results would be reported as they come in (streaming).

We currently depend on `async-http-client-backend-cats` but that should be changed to `async-http-client-backend-fs2` for streaming data

**Comment by Mobrockers** [comment]
This is something I think would fit better as a feature of the dashboard. That way we only have to implement it once, and adds an extra use case for using the dashboard.

This is also how other services like sonarcloud work.

**Comment by Durveshpal** [comment]
I think you give an extra line after the comment statement and  would fit better as a feature of the dashboard, where you can configure the dashboard to post the result to GitHub as a status check

**Comment by cbrunnkvist** [comment]
As @Mobrockers points out, if submitting the _check_ instead is a responsibility of the dashboard, then you can avoid having one GitHub API client to this in _stryker4s_, and another one in _stryker-net_ and so on.

There are a couple of other things to put on the wish list here: reporting _changes_ in score between builds/checks is a major one. Features like that (tracking reports by commit rather then by branch) also fit much better in the dashboard project.

**Comment by hugo-vrijswijk** [comment]
I agree it does make more sense for this functionality to be in the Stryker dashboard. I'll move this issue there and edit the description
